### PR TITLE
efi: Improve Directory Open/Read

### PR DESCRIPTION
**Pull request by retrage**
This PR mainly proposes two features:
1. 3449aa1: Add open from non-root directory support
2. c43e59d: Add directory read support

Some bootloaders (e.g., `fbx64.efi` from shim) need these features to traverse the given filesystem. The rough design is based on the discussion in issue #19, which introduces `Node` enum to abstract `Directory` and `File`.

It also includes some small fixes for EFI compatibility.

Once this PR is merged, issue #19 can be closed.

### CODE REVIEW

File: src/common.rs [hunk 1]
```diff
@@ -64,6 +64,17 @@ pub fn ucs2_as_ascii_length(input: *const u16) -> usize {
...
+pub fn ascii_length(input: &str) -> usize {
+    let mut len = 0;
+    for c in input.chars() {
+        if c == '\0' {
+            break;
+        }
+        len += 1;
+    }
+    len
+}
```
Comment (rbradford) [inline src/common.rs:76]: Sort of surprised you need to need to do this, is there not a core::str::len ?

**Review by rbradford — APPROVED** [review]
Cool! Does this now allow us to boot the stock Ubuntu images?

File: src/fat.rs [hunk 2]
```diff
@@ -245,6 +343,37 @@ impl<'a> Directory<'a> {
...
+            FileType::File => Ok((
+                self.filesystem
+                    .get_file(de.cluster, de.size)
+                    .unwrap()
+                    .into(),
+                name,
+            )),
+        }
+    }
+    pub fn open(&self, path: &str) -> Result<Node, Error> {
```
Comment (rbradford) [inline src/fat.rs:366]: nit: Line of whitespace here

File: src/common.rs (hunk 1 above)
Comment (rbradford) [reply src/common.rs:76]: Or is this slightly different? This is more like `strlen()`?

**Review by rbradford — APPROVED** [review]

File: src/common.rs (hunk 1 above)
Comment (retrage) [reply src/common.rs:76]: Actually, `core::str::len()` returns size of the array when given str is converted from [u8; N].

Let's run the following snippet.
```rust
fn main() {
    let s = "foo";
    let mut a = [0_u8; 8];
    a[0] = b'b';
    a[1] = b'a';
    a[2] = b'r';
    let u = core::str::from_utf8(&a).unwrap();
    println!("s.len(): {}", s.len());
    println!("u.len(): {}", u.len());
    println!("ascii_length(u): {}", ascii_length(u));
}
```
We got:

```
s.len(): 3
u.len(): 8
ascii_length(u): 3
```

`s` is string literal, and len() returns 3 as expected. But `u` type is str too, but it is from [u8; N], and len() returns 8. This is different behavior from C's `strlen()`.

Note: We can avoid this pitfall by calling `u.trim_matches(char::from(0));` before `len()` call. I'll update to use this.

File: src/fat.rs (hunk 2 above)
Comment (retrage) [reply src/fat.rs:366]: Thanks, I'll fix it.

**Comment by retrage** [comment]
> Cool! Does this now allow us to boot the stock Ubuntu images?

Not yet. I'm working on it in another branch:
https://github.com/retrage/rust-hypervisor-firmware/tree/vanilla-ubuntu-boot

It seems to require more EFI compatibilities like `load_image()` and `start_image()` support.

File: src/fat.rs (hunk 2 above)
Comment (retrage) [reply src/fat.rs:367]: Fixed.

**Pull request merged**